src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -175,6 +181,301 @@ static inline ASR::asr_t* create_UnaryFunction(Allocator& al, const Location& lo
+                    EXPR(ASR::make_ListItem_t(al, loc, lps, i, int_type, nullptr)),
+                    pi_len, nullptr)));
+                if_body_1.push_back(al, STMT(ASR::make_Assignment_t(al, loc, i,
+                    EXPR(ASR::make_IntegerBinOp_t(al, loc, i, ASR::binopType::Add,
+                    EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                    int_type, nullptr)), nullptr)));
+                Vec<ASR::stmt_t *> else_body_1; else_body_1.reserve(al, 1);
+                {
+                    Vec<ASR::stmt_t *> if_body_2; if_body_2.reserve(al, 1);
+                    Vec<ASR::stmt_t *> else_body_2; else_body_2.reserve(al, 1);
+                    ASR::expr_t *a_test = EXPR(ASR::make_IntegerCompare_t(al, loc,
+                        pi_len, ASR::cmpopType::NotEq,
+                        EXPR(ASR::make_IntegerConstant_t(al, loc, 0, int_type)),
+                        logical_type, nullptr));
+                    if_body_2.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                        pi_len, EXPR(ASR::make_ListItem_t(al, loc, lps,
+                            EXPR(ASR::make_IntegerBinOp_t(al, loc, pi_len,
+                                ASR::binopType::Sub,
+                                EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                                int_type, nullptr)),
+                            int_type, nullptr)), nullptr)));
+                    else_body_2.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                        i, EXPR(ASR::make_IntegerBinOp_t(al, loc, i,
+                            ASR::binopType::Add,
+                            EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                            int_type, nullptr)), nullptr)));
+                    else_body_1.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
+                        if_body_2.p, if_body_2.n, else_body_2.p, else_body_2.n)));
+                }
+                loop_body.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
+                    if_body_1.p, if_body_1.n, else_body_1.p, else_body_1.n)));
+            }
+
+            body.push_back(al, STMT(ASR::make_WhileLoop_t(al, loc, nullptr, a_test,
+                loop_body.p, loop_body.n)));
+        }
+        body.push_back(al, STMT(ASR::make_Assignment_t(al, loc, j, EXPR(
+            ASR::make_IntegerConstant_t(al, loc, 0, int_type)), nullptr)));
+        body.push_back(al, STMT(ASR::make_Assignment_t(al, loc, i, EXPR(
+            ASR::make_IntegerConstant_t(al, loc, 0, int_type)), nullptr)));
+        {
+            ASR::expr_t *a_test = EXPR(ASR::make_LogicalBinOp_t(al, loc,
+                EXPR(ASR::make_IntegerCompare_t(al, loc,
+                    EXPR(ASR::make_IntegerBinOp_t(al, loc, s_len,
+                        ASR::binopType::Sub, i, int_type, nullptr)),
+                    ASR::cmpopType::GtE,
+                    EXPR(ASR::make_IntegerBinOp_t(al, loc, pat_len,
```

Comment:
This can be simplified as well. See `create_ElementalAdd` and `create_ElementalBinOp` in https://github.com/lfortran/lfortran/pull/1577.